csharp/src/Ice/ITransceiver.cs

```diff
@@ -11,302 +11,52 @@
-        {
-            return await Read(this, buffer, offset, partial).ConfigureAwait(false);
-
-            static async Task<ArraySegmentAndOffset> Read(ITransceiver self, ArraySegment<byte> buffer, int offset,
-                bool partial)
-            {
-                var result = new TaskCompletionSource<ArraySegmentAndOffset>();
-                var p = new ArraySegmentAndOffset(buffer, offset);
-                async void ReadCallback(object state)
-                {
-                    try
-                    {
-                        var transceiver = (ITransceiver)state;
-                        int status;
-                        lock (transceiver)
-                        {
-                            transceiver.FinishRead(ref p.Buffer, ref p.Offset);
-                            status = transceiver.Read(ref p.Buffer, ref p.Offset);
-                        }
-                        if ((status & SocketOperation.Write) != 0)
-                        {
-                            await transceiver.WriteImplAsync(
-                                new List<ArraySegment<byte>>(), 0, partial).ConfigureAwait(false);
-                        }
-                        if ((status & SocketOperation.Read) != 0 && !partial)
-                        {
-                            result.SetResult(await transceiver.ReadImplAsync(p.Buffer, p.Offset,
-                                false).ConfigureAwait(false));
-                        }
-                        else
-                        {
-                            result.SetResult(p);
-                        }
-                    }
-                    catch (System.Exception ex)
-                    {
-                        result.SetException(ex);
-                    }
-                }
-
-                lock (self)
-                {
-                    if (self.StartRead(ref p.Buffer, ref p.Offset, ReadCallback, self))
-                    {
-                        ReadCallback(self);
-                    }
-                }
-                return await result.Task.ConfigureAwait(false);
-            }
-        }
+        /// <param name="size">The size of the message to check.</param>
+        // TODO: Remove this? This is used to ensure the user doesn't try to send a message which is larger
+        // than the datagram size... UDP users would be better of not sending large datagrams instead of us
+        // trying to softly error if the user sends a large datagram.
+        void CheckSendSize(int size);
 
+        /// <summary>Closes the transceiver. The transceiver might use this method to send a notification to the peer
+        /// of the connection closure.</summary>
+        /// <param name="exception">The reason of the connection closure.</param>
+        /// <param name="cancel">A cancellation token that receives the cancellation requests.</param>
+        ValueTask ClosingAsync(Exception exception, CancellationToken cancel);
```

Comment:
I thought ClosingAsync was more appropriate as we are initiating a connection closure here... I'll rename since you and José find CloseAsync more appropriate...